MWPW-144549 CTA alignment for Text, Icon, and Media Blocks

meganthecoder opened the pull request:
* Align CTAs of Text, Icon, and Media blocks in "-up" sections

Resolves: [MWPW-144549](https://jira.corp.adobe.com/browse/MWPW-144549)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/drafts/methomas/text-alignment
- After: https://methomas-text-alignment--milo--adobecom.hlx.page/drafts/methomas/text-alignment?martech=off

Bacom:
- Before: https://main--bacom--adobecom.hlx.live/products/experience-manager/assets/pricing
- After: https://main--bacom--adobecom.hlx.live/products/experience-manager/assets/pricing?milolibs=methomas-text-alignment
- Before: https://main--bacom--adobecom.hlx.live/
- After: https://main--bacom--adobecom.hlx.live/?milolibs=methomas-text-alignment

Kitchen sink:
- Text: https://methomas-text-alignment--milo--adobecom.hlx.page/docs/library/kitchen-sink/text
- Icon: https://methomas-text-alignment--milo--adobecom.hlx.page/docs/library/kitchen-sink/icon-block
- Media: https://methomas-text-alignment--milo--adobecom.hlx.page/docs/library/kitchen-sink/media

Edge cases and regression:
- https://methomas-text-alignment--milo--adobecom.hlx.page/drafts/methomas/text-alignment-all

JasonHowellSlavin reviewed (approved):
LGTM.

Really like that you included Kitchen Sink for verification!

meganthecoder commented:
I'm closing this pull request for now because I've found some edge cases to account for. I'll re-open when it's ready to look at again. Thanks!

meganthecoder commented:
I've reopened the PR, and it's ready for review!
Here's a lengthier test page of mostly edge cases. I added all variants, even ones that are unlikely to be used in the "-up" pattern: https://methomas-text-alignment--milo--adobecom.hlx.page/drafts/methomas/text-alignment-all

JasonHowellSlavin reviewed (commented):
One comment, looks good otherwise!

elan-tbx reviewed (approved):
Looks good to me, thanks for doing this!

ryanmparrish commented on libs/blocks/icon-block/icon-block.js:
```diff
@@ -75,6 +75,17 @@ function decorateContent(el) {
       if (secondColumn.children.length === 1) el.classList.add('items-center');
       el.querySelector('.foreground .text-content').append(secondColumn);
     }
+    const ctas = el.querySelectorAll('.action-area');
```
These are div.action-area(s) right? Maybe update to `const actionAreas...`?  
Is there potential here for multiples. On the selectors in media.js and text.js you are only getting the single last-of-type. Im just trying to understand why this block may have multiple action areas where the others are assumed to have just one.

meganthecoder replied:
Sure I can update the variable name.
Here's the use case:
<img width="1473" alt="Screenshot 2024-04-05 at 3 58 03 PM" src="https://github.com/adobecom/milo/assets/19690988/b9cdbba0-94ff-4935-90d7-63866243a532">
From https://main--milo--adobecom.hlx.page/drafts/methomas/text-alignment-all
It's kind of strange. In the icon block, we're adding an `action-area` class around the standalone link: https://github.com/adobecom/milo/blob/51a113452fccf258fdd91d6edfba348775a4ba25/libs/blocks/icon-block/icon-block.js#L63
So if there's a button and a standalone link, they're in separate action areas. Text and media blocks don't do this. If those blocks have links next to each other, they all get added to the same action area by the button decorator.

ryanmparrish reviewed (commented):
Looking good. Thanks. I left one question in the `icon-block.js` file. 
Also Is there any layout considerations for when/if a `.supplemental-text` or `subtext` is used after the .action-area. I made a draft [here](https://methomas-text-alignment--milo--adobecom.hlx.page/drafts/rparrish/text-alignment) to explore these and for the most part it looks to just ignore those use cases since the .action-area is no longer the :last-child etc. Just wondering if that was a consideration or concern?

meganthecoder commented:
> Also Is there any layout considerations for when/if a `.supplemental-text` or `subtext` is used after the .action-area. I made a draft [here](https://methomas-text-alignment--milo--adobecom.hlx.page/drafts/rparrish/text-alignment) to explore these and for the most part it looks to just ignore those use cases since the .action-area is no longer the :last-child etc. Just wondering if that was a consideration or concern?

@ryanmparrish I'll take a look and see if there's anything to clean up. Are these just edge cases or are there real use cases where we support this?

Ruchika4 reviewed (approved):
Looks good.

meganthecoder commented:
> Also Is there any layout considerations for when/if a `.supplemental-text` or `subtext` is used after the .action-area. I made a draft [here](https://methomas-text-alignment--milo--adobecom.hlx.page/drafts/rparrish/text-alignment) to explore these and for the most part it looks to just ignore those use cases since the .action-area is no longer the :last-child etc. Just wondering if that was a consideration or concern?

@ryanmparrish There's a use case for the media block that has text after the action areas so the solution was to exclude the alignment functionality in that case. 

<img width="609" alt="Screenshot 2024-04-10 at 4 21 54 PM" src="https://github.com/adobecom/milo/assets/19690988/72e47aa4-aef1-403e-ab8c-a9ee88ff731a">

That's done with the `:last-of-type` selector, which I had on the media and text blocks. I've rewritten the icon block functionality to match better so they all work that way.

I've also added your use cases to my edge case test page: https://methomas-text-alignment--milo--adobecom.hlx.page/drafts/methomas/text-alignment-all

robert-bogos commented:
Hi @meganthecoder ! I see that the unit tests check is failing, can you please look into that?

meganthecoder commented:
@robert-bogos The unit tests are passing now, so I'm adding back the Ready for Stage label. In any case, the failed unit tests were for bulk-publish-v2 and unrelated to this PR.

meganthecoder commented:
Looks like this won't get out today, so I'll put it on hold until I get back from PTO
